### PR TITLE
Add disk dialog bug fixes

### LINKED ISSUE #1272: AddDiskModalBody does not properly initialize its `format` state
```
> error: Warning: Failed %s type: %s%s prop The prop `format` is marked as required in `VolumeCreateBody`, but its value is `undefined`. 
    at VolumeCreateBody (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:130756:5)
    at CreateNewDisk (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:133599:5)
    at form
    at FormBase (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:121187:11)
    at Form
    at div
    at ModalBoxBody (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:121349:11)
    at div
    at ModalBox (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:121385:11)
    at div
    at FocusTrap2 (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:113655:11)
    at div
    at Backdrop (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:121340:11)
    at ModalContent (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:121464:11)
    at _Modal (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:121529:7)
    at AddDiskModalBody (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:133718:7)
    at WithDialogs (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:120550:24)
    at VmDetailsPage (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:161086:5)
    at AppActive (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:163583:7)
    at App (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:163531:81)
```

[This error](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1270-20231026-134109-32e5fa30-fedora-38-devel/log.html#24) is really awkward. `AddDiskModalBody` has a very async way of initializing its `this.state.format`, via `componentDidMount` → `Promise` → `initialState`. So the initial render() will always run with it undefined. But also, `getDefaultVolumeFormat()` can easily return `undefined`.  Fixing this requires someone with knowledge how this should be fixed *properly*. I tried a hack like this:
```diff
--- src/components/vm/disks/diskAdd.jsx
+++ src/components/vm/disks/diskAdd.jsx
@@ -377,6 +377,7 @@ export class AddDiskModalBody extends React.Component {
             customDiskVerificationMessage: null,
             verificationInProgress: false,
             file: null,
+            format: undefined,
         };
         this.onValueChanged = this.onValueChanged.bind(this);
         this.dialogErrorSet = this.dialogErrorSet.bind(this);
@@ -399,6 +400,7 @@ export class AddDiskModalBody extends React.Component {
                     .map(pool => ({ name: pool.name, type: pool.type }))
                     .sort(sortFunction)[0];
 
+        console.log("XXX AddDiskModalBody initialState; defaultPool", defaultPool, "format", defaultPool ? getDefaultVolumeFormat(defaultPool) : "none");
         return {
             storagePools,
             vm,
@@ -412,7 +414,7 @@ export class AddDiskModalBody extends React.Component {
             existingVolumeName: undefined,
             size: 1,
             unit: units.GiB.name,
-            format: defaultPool && getDefaultVolumeFormat(defaultPool),
+            format: defaultPool ? getDefaultVolumeFormat(defaultPool) : "none",
             target: availableTarget,
             permanent: !domainIsRunning(vm.state), // default true for a down VM; for a running domain, the disk is attached tentatively only
             hotplug: domainIsRunning(vm.state), // must be kept false for a down VM; the value is not being changed by user
@@ -732,6 +734,9 @@ export class AddDiskModalBody extends React.Component {
     }
 
     render() {
+        if (this.state.format === undefined) // componentDidMount → initialState did not run yet
+            return;
+
         const Dialogs = this.context;
         const { dialogLoading, vm, storagePools, vms } = this.state;
         const { isMediaInsertion } = this.props;

```

which helps a bit, but breaks the test.

This React warning is really useful and points out an actual bug.

_Originally posted by @martinpitt in https://github.com/cockpit-project/cockpit-machines/issues/1270#issuecomment-1782337326_

### LINKED ISSUE #1273: AddDiskModalBody passes invalid size to VolumeCreateBody
This is exposed by [these React errors](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1270-20231027-062752-7b6401ac-fedora-38-devel/log.html) (all four tests fail on the same issue):
```
error: Warning: Failed %s type: %s%s prop Invalid prop `size` of type `string` supplied to `VolumeCreateBody`, expected `number`. 
    at VolumeCreateBody (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:130756:5)
    at CreateNewDisk (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:133599:5)
   [...]
    at AddDiskModalBody (http://127.0.0.2:9091/cockpit/@localhost/machines/index.js:133718:7)
```

This is a real bug. In most cases, the passed property is the default from `initState`:
```
> log: XXX AddDiskModalBody.render size {"type":"number","value":1,"description":"1"}
```
but during testAddDiskDirPool() it becomes:
```
> log: XXX AddDiskModalBody.render size 2
```
i.e. a string. It's not really clear to me where this gets set, there is no direct setState() call.

I'll keep ignoring this specific log message for now, but this should be fixed. In particular, `AddDiskModalBody` also has a state `unit`, which is entirely ignored when passing the size to VolumeCreateBody.

### PR DESCRIPTION
**Pull request by martinpitt**
This has turned into quite a can of worms... This has just renewed my aversion to abusing `useEffect()` -- honestly they should *all* die in diskAdd.jsx, I did not see any valid use case of it. They look like sloppy "let's clean up after our own  bad/wrong decisions", and make state handling immensely brittle and hard to follow. The logic from all of them should be moved into a single `useInit()` to set up the dialog, and the corresponding event handlers (which ought to set up the state correctly right away instead of relying on effects to second-guess them).

This PR at least gets rid of two of them.

Fixes #1272

Fixes #1273

### CODE REVIEW

**Comment by martinpitt** [comment]
This [spotted some actual bugs](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1327-20231124-073522-8b58bc6c-fedora-39-devel/log.html#24) -- the `validVolumeFormats` computation is wrong. E.g. for an "fs" pool there are [various Linux file systems](https://libvirt.org/storage.html#filesystem-pool) as valid options, but most certainly not "qcow"/"raw". Looking..

File: src/components/storagePools/storageVolumeCreateBody.jsx [hunk 1]
```diff
@@ -65,15 +65,20 @@ const VolumeDetails = ({ idPrefix, size, unit, format, storagePoolCapacity, stor
...
-            <FormGroup fieldId={`${idPrefix}-fileformat`} label={_("Format")}>
-                <FormSelect id={`${idPrefix}-format`}
-                    onChange={(_event, value) => onValueChanged('format', value)}
-                    value={format}>
-                    { validVolumeFormats.map(format => <FormSelectOption value={format} key={format} label={format} />) }
-                </FormSelect>
-            </FormGroup>
-        );
+        if (!format)
+            console.error("VolumeDetails internal error: format is not set for storage pool type", storagePoolType);
```
Comment (cockpituous) [inline src/components/storagePools/storageVolumeCreateBody.jsx:69]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1327-20231124-094238-e70920c2-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

Comment (martinpitt) [reply src/components/storagePools/storageVolumeCreateBody.jsx:71]: Annotated locally, will push with the next round.

File: src/components/storagePools/storageVolumeCreateBody.jsx [hunk 2]
```diff
@@ -62,18 +63,24 @@ const VolumeDetails = ({ idPrefix, size, unit, format, storagePoolCapacity, stor
...
```
Comment (cockpituous) [inline src/components/storagePools/storageVolumeCreateBody.jsx:71]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1327-20231124-101203-c40ec8ba-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

Comment (cockpituous) [inline src/components/storagePools/storageVolumeCreateBody.jsx:71]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1327-20231124-101157-c40ec8ba-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
Argh, that last commit 3e110b22ac6df8d7d831now breaks [testAddDiskDirPool again](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1327-20231124-130818-3e110b22-arch/log.html#22-2). That commit *should* be completely innocuous -- it makes state updates more synchronous.

Honestly, I cannot believe how brittle all of this is -- whenever I touched *anything* in the code to fix one bug, the maze of useEffect()s breaks something else -- that's what I meant with "impossible to track state changes". The dialog is hanging forever or this wrong format.

**Comment by martinpitt** [comment]
@KKoukiou Please be picky here -- I'm not patient/experienced enough to fix all of it in one go, but it should at least go into the right direction. Please let me know if you don't have time, then I'll ask somebody else (Jelle most likely) to review. Epharisto!

**Comment by martinpitt** [comment]
I added a fix for #1273, which was fortunately quite easy now.

**Comment by martinpitt** [comment]
Ah, the size data type mismatch [affected the "Create storage volume" dialog as well](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1327-20231127-081517-6f2c5ee1-fedora-39-devel/log.html#71), fixed.
